ospfd/ospf_sr.c

```diff
@@ -636,70 +618,39 @@ static int ospf_zebra_send_mpls_labels(int cmd, struct sr_nhlfe nhlfe)
 			   inet_ntoa(nhlfe.prefv4.prefix),
 			   nhlfe.prefv4.prefixlen, nhlfe.ifindex);
 
-	return zclient_send_message(zclient);
-}
-
-/* Request zebra to install/remove FEC in FIB */
-static int ospf_zebra_send_mpls_ftn(int cmd, struct sr_nhlfe nhlfe)
-{
-	struct zapi_route api;
-	struct zapi_nexthop *api_nh;
-
-	/* Support only IPv4 */
-	if (nhlfe.prefv4.family != AF_INET)
-		return -1;
-
-	memset(&api, 0, sizeof(api));
-	api.vrf_id = VRF_DEFAULT;
-	api.type = ZEBRA_ROUTE_OSPF;
-	api.safi = SAFI_UNICAST;
-	memcpy(&api.prefix, &nhlfe.prefv4, sizeof(struct prefix_ipv4));
-
-	if (cmd == ZEBRA_ROUTE_ADD) {
-		/* Metric value. */
-		SET_FLAG(api.message, ZAPI_MESSAGE_METRIC);
-		api.metric = OSPF_SR_DEFAULT_METRIC;
-		/* Nexthop */
-		SET_FLAG(api.message, ZAPI_MESSAGE_NEXTHOP);
-		api_nh = &api.nexthops[0];
-		IPV4_ADDR_COPY(&api_nh->gate.ipv4, &nhlfe.nexthop);
-		api_nh->type = NEXTHOP_TYPE_IPV4_IFINDEX;
-		api_nh->ifindex = nhlfe.ifindex;
-		/* MPLS labels */
-		SET_FLAG(api.message, ZAPI_MESSAGE_LABEL);
-		api_nh->labels[0] = nhlfe.label_out;
-		api_nh->label_num = 1;
-		api_nh->vrf_id = VRF_DEFAULT;
-		api.nexthop_num = 1;
-	}
-
-	if (IS_DEBUG_OSPF_SR)
-		zlog_debug("    |-  %s FEC %u for %s/%u via %u",
-			   cmd == ZEBRA_ROUTE_ADD ? "Add" : "Delete",
-			   nhlfe.label_out, inet_ntoa(nhlfe.prefv4.prefix),
-			   nhlfe.prefv4.prefixlen, nhlfe.ifindex);
-
-	return zclient_route_send(cmd, zclient, &api);
+	zl.type = ZEBRA_LSP_OSPF_SR;
+	zl.local_label = nhlfe.label_in;
+
+	SET_FLAG(zl.message, ZAPI_LABELS_FTN);
+	zl.route.prefix.family = nhlfe.prefv4.family;
+	zl.route.prefix.prefixlen = nhlfe.prefv4.prefixlen;
+	zl.route.prefix.u.prefix4 = nhlfe.prefv4.prefix;
+	zl.route.type = ZEBRA_ROUTE_OSPF;
+	zl.route.instance = 0;
+
+	zl.nexthop_num = 1;
+	znh = &zl.nexthops[0];
+	znh->type = NEXTHOP_TYPE_IPV4_IFINDEX;
+	znh->family = AF_INET;
+	znh->address.ipv4 = nhlfe.nexthop;
+	znh->ifindex = nhlfe.ifindex;
+	znh->label = nhlfe.label_out;
+
```

Comment:
That's not good. You need to add one more test to check if the nhlfe.label_out is not equal to MPLS_LABEL_IMPLICIT_NULL because you remove the test in add_sid_nhlfe() and del_sid_nhlfe() functions.

Reply:
OK. I miss the fact that we need entry even if the label is MPLS_LABEL_IMPLICIT_NULL.

The only remaining problem concerns the Node-SID with NO-PHP-FLAG. The code will add entry to POP label and forward packet to IP address of loopback interface instead of forward to loopback interface only. A test on that could be inserted in order to set `znh->type` to NEXTHOP_TYPE_IF_INDEX instead of NEXTHOP_TYPE_IPV4_IFINDEX. But, the old code suffer from the same problem.

I'll solve this as well as ECMP in another patch